monasca_rsyslog/client.py

```diff
@@ -46,13 +60,83 @@ def __init__(self):
-            headers={ 'Content-Type': 'application/json' },
-            data=data
-        )
+    def _combine_logs(self, data, log_buffer):
+        """ Combine with the existing log buffer if not empty. """
+
+        log_count = 0
+        if data != None:
+            if self._verbosity > 1:
+                print(data)
+                sys.stdout.flush()
+            for key, value in jsonutils.loads(data).items():
+                log_buffer.setdefault(key, []).extend(value)
+                log_count += len(value)
+        else:
+            if self._verbosity > 1:
+                print("No input from stdin for {} seconds.".format(self._min_poll_delay))
+                sys.stdout.flush()
+        return log_count
+
+    def _retry_post_logs(self, log_buffer):
+        """ Try to post logs forever until success. """
+
+        while True:
+            try:
+                # Try to post the things in the buffer
+                self._sess.post(
+                    '/logs',
+                    endpoint_override=self._url,
+                    headers={ 'Content-Type': 'application/json' },
+                    data=jsonutils.dumps(log_buffer)
+                )
+                break
+            except Exception as e:
+                # If there is a failure, most likely ConnectionFailure,
+                # write to stdout and stderr then try again.
+                sys.stderr.write("Connection failure, trying again.")
+                sys.stderr.write(e)
+                sys.stderr.flush()
+                time.sleep(1)
+
+    def _log_generator(self, log_source, poll_interval):
+        """ Helper for performing line-by-line reads of input source. """
+
+        while True:
+            buffer_is_not_empty, _, _ = select([log_source], [], [], poll_interval)
+            if buffer_is_not_empty:
+                line = log_source.readline()
+                if line:
+                    yield line
+                else:
+                    return
+            else:
+                # Waiting for stdin buffer timed out, yeild in case there are
+                # things waiting to be flushed but no input for timeout duration.
+                yield None
+
+    def handle_logs(self, log_source=sys.stdin):
+        """ Post logs to Monasca which are suitably pre-encoded as JSON. """
+
+        # Initialise the variables
+        log_count, log_buffer, start_time = 0, {}, time.time()
+
+        # Iterate through the stdin function
+        for data in self._log_generator(log_source=log_source, poll_interval=self._min_poll_delay):
+            log_count += self._combine_logs(data, log_buffer)
+            elapsed_time = time.time() - start_time
+            waited_too_long = elapsed_time > self._min_poll_delay
+            buffer_too_long = log_count > self._max_batch_size
+            if (waited_too_long or buffer_too_long) and log_count > 0:
+                # Print a summary of states
+                if self._verbosity > 0:
```

Comment:
nit: This stanza could sit in a function since we use it multiple times (or use an external logging library), but not to worry.

Reply:
Good idea, done!